tablecodec/tablecodec.go

```diff
@@ -729,6 +755,73 @@ func decodeRestoredValues(columns []rowcodec.ColInfo, restoredVal []byte) ([][]b
 	return resultValues, nil
 }
 
+// decodeRestoredValuesV5 decodes index values whose format is introduced in TiDB 5.0.
+// Unlike the format in TiDB 4.0, the new format is optimized for storage space:
+// 1. If the index is a composed index, only the non-binary string column's value need to write to value, not all.
+// 2. If a string column's collation is _bin, then we only write the number of the truncated spaces to value.
+// 3. If a string column is char, not varchar, then we use the sortKey directly.
+func decodeRestoredValuesV5(columns []rowcodec.ColInfo, keyVal [][]byte, restoredVal []byte) ([][]byte, error) {
+	colIDs := make(map[int64]int, len(columns))
+	result := make([][]byte, len(columns))
+	// restoredData2All is the map from the offset in restoredColumns to the offset in columns.
+	restoredData2All := make(map[int]int)
+	restoredColumns := make([]rowcodec.ColInfo, 0, len(columns))
+	j := 0
+
+	// Collate some information, restoredColumns means the columns whose value need to restore from the index value.
+	for i, col := range columns {
+		if types.NeedRestoredData(col.Ft) {
+			colIDs[col.ID] = j
+			restoredData2All[j] = i
+			j++
+			copyColInfo := rowcodec.ColInfo{
+				ID: col.ID,
+				Ft: columns[i].Ft.Clone(),
```

Comment:
We don't need to clone the field type.

Reply:
Otherwise, it would be modified.

Reply:
It's a pointer, we just change the pointer value, the content if the type is not changed.